src/Popup.js

```diff
@@ -11,45 +11,54 @@ export default class Popup extends MapComponent {
 
   componentWillMount() {
     super.componentWillMount();
-    const { children, map, ...props } = this.props;
-    this.leafletElement = popup(props);
+    const { children, map, popupContainer, ...props } = this.props;
+    this.leafletElement = popup(props, popupContainer);
+    this.leafletElement.on('open', () => { //on popoup create new react root
+      const {children} = this.props
+      this.contentComponent = React.render(React.DOM.div({children}), this.leafletElement._contentNode)
+      this.leafletElement._adjustPan()
+    }, this)
+    this.leafletElement.on('close', () => { //remove react root
+      delete this.contentComponent
+      React.unmountComponentAtNode(this.leafletElement._contentNode)
+    }, this)
+  }
+
+  componentDidMount() {
+    const { popupContainer, map, position } = this.props;
+    super.componentDidMount();
+    //Attach to container component
+    if (popupContainer) {
+      popupContainer.bindPopup(this.leafletElement)
+    } else { //attach to map
+      if(position) this.leafletElement.setLatLng(position)
+      this.leafletElement.openOn(map)
+    }
   }
 
   componentDidUpdate(prevProps) {
-    const { children, map, popupContainer, position, ...props } = this.props;
-    if (children !== prevProps.children) {
-      const content = React.renderToStaticMarkup(children);
-      if (popupContainer) {
-        popupContainer.bindPopup(content, props);
-      }
-      else {
-        const el = this.leafletElement;
-        el.setContent(content);
-        if (position !== prevProps.position) el.setLatLng(position);
-      }
+    const { children, popupContainer, position } = this.props;
+    if(!popupContainer) {
```

Comment:
Sure. I use different style when I write my other code so its easy to miss for me.